modules/core/src/main/java/org/apache/ignite/internal/processors/cache/persistence/defragmentation/TreeIterator.java

```diff
@@ -0,0 +1,149 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.cache.persistence.defragmentation;
+
+import java.nio.ByteBuffer;
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.internal.processors.cache.persistence.pagemem.PageMemoryEx;
+import org.apache.ignite.internal.processors.cache.persistence.tree.BPlusTree;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusLeafIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.BPlusMetaIO;
+import org.apache.ignite.internal.processors.cache.persistence.tree.io.PageIO;
+import org.apache.ignite.internal.util.GridUnsafe;
+
+/** */
+public class TreeIterator {
+    /** Direct memory buffer with a size of one page. */
+    private final ByteBuffer pageBuf;
+
+    /** Offheap page size. */
+    private final int pageSize;
+
+    /** */
+    public TreeIterator(int size) {
+        pageSize = size;
+
+        pageBuf = ByteBuffer.allocateDirect(pageSize);
+    }
+
+    // Performance impact of constant closures allocation is not clear. So this method should be avoided in massive
+    // operations like tree leaves access.
+    /** */
+    public static <T> T access(
+        PageAccessType access,
+        PageMemoryEx pageMemory,
+        int grpId,
+        long pageId,
+        PageAccessor<T> accessor
+    ) throws IgniteCheckedException {
+        assert access != null;
+        long page = pageMemory.acquirePage(grpId, pageId);
+
+        try {
+            long pageAddr = access == PageAccessType.ACCESS_READ
+                    ? pageMemory.readLock(grpId, pageId, page)
+                    : pageMemory.writeLock(grpId, pageId, page);
+
+            try {
+                return accessor.access(pageAddr);
+            }
+            finally {
+                if (access == PageAccessType.ACCESS_READ)
+                    pageMemory.readUnlock(grpId, pageId, page);
+                else
+                    pageMemory.writeUnlock(grpId, pageId, page, null, true);
+            }
+        }
+        finally {
+            pageMemory.releasePage(grpId, pageId, page);
+        }
+    }
+
+    /** */
+    @SuppressWarnings("PublicInnerClass")
+    public enum PageAccessType {
```

Comment:
Let's remove such customization and create simple read/write methods.

Reply:
You know, "access(ACCESS_READ, partCtx.cachePageMemory, partCtx.grpId, partMetaPageId, ..." looks better then
"access(false, partCtx.cachePageMemory, partCtx.grpId, partMetaPageId, "

That was the reason. I can create constants like "boolean ACCESS_READ = false;", but strict typing works better here I think.